### PR TITLE
Clamp border corner radius

### LINKED ISSUE #2183: Border rendering glitches
Bad (taken from wrench with dbg alpha primitives / render target on):

Good (taken from a jsfiddle I made [here](https://jsfiddle.net/LLg2arf9/)):

```
---
root: 
  items: 
    - 
      bounds: [0, 0, 1024, 768]
      "clip-rect": [0, 0, 1024, 768]
      "clip-and-scroll": 0
      "backface-visible": true
      type: "stacking-context"
      "scroll-policy": scrollable
      "transform-style": flat
      filters: []
      items: 
        - 
          bounds: [0, 0, 200, 200]
          "clip-rect": [0, 0, 200, 200]
          "clip-and-scroll": 0
          "backface-visible": true
          type: border
          width: 10
          "border-type": normal
          color: red
          style: dashed
          radius: [180, 180]
        - 
          bounds: [200, 0, 380, 380]
          "clip-rect": [200, 0, 380, 380]
          "clip-and-scroll": 0
          "backface-visible": true
          type: border
          width: 10
          "border-type": normal
          color: green
          style: dashed
          radius: [180, 180]
        - 
          bounds: [580, 0, 200, 200]
          "clip-rect": [580, 0, 200, 200]
          "clip-and-scroll": 0
          "backface-visible": true
          type: border
          width: 10
          "border-type": normal
          color: red  
          style: solid
          radius: [180, 180]
        - 
          bounds: [780, 0, 380, 380]
          "clip-rect": [780, 0, 380, 380]
          "clip-and-scroll": 0
          "backface-visible": true
          type: border
          width: 10
          "border-type": normal
          color: green
          style: solid
          radius: [180, 180]
        - 
          bounds: [0, 380, 200, 200]
          "clip-rect": [0, 380, 200, 200]
          "clip-and-scroll": 0
          "backface-visible": true
          type: border
          width: 10
          "border-type": normal
          color: red
          style: dotted
          radius: [180, 180]
        - 
          bounds: [200, 380, 380, 380]
          "clip-rect": [200, 380, 380, 380]
          "clip-and-scroll": 0
          "backface-visible": true
          type: border
          width: 10
          "border-type": normal
          color: green
          style: dotted
          radius: [180, 180]
  id: [0, 0]
pipelines: []
```

From what I can tell it hasto do with how the bounds of the border relate to the border radius. You'll notice in the green examples the only difference is the bounds are bigger.

### PR DESCRIPTION
**Pull request by nc4rrillo**
Fixes #2183 

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/2187)

### CODE REVIEW

**Comment by mstange** [comment]
This is a bit too restrictive - corner radii of more than half the width or height are permitted in CSS if the neighboring corner leaves enough space. See the testcase in #2174 for an example.

Here's the code in Firefox that limits corner radii: https://searchfox.org/mozilla-central/rev/f5f1c3f294f89cfd242c3af9eb2c40d19d5e04e7/layout/generic/nsFrame.cpp#1699-1720
The radii of neighboring corners are scaled down proportionally.  The spec text is at https://drafts.csswg.org/css-backgrounds/#corner-overlap .

**Comment by nc4rrillo** [comment]
@mstange I've taken another stab at it, I'm not experienced at reading the specs but I _think_ this matches the Gecko functionality.

I get


with the following YAML which incorporates your case in #2174:

```
---
root: 
  items: 
    - 
      bounds: [0, 0, 1024, 768]
      "clip-rect": [0, 0, 1024, 768]
      "clip-and-scroll": 0
      "backface-visible": true
      type: "stacking-context"
      "scroll-policy": scrollable
      "transform-style": flat
      filters: []
      items: 
        - 
          bounds: [0, 0, 200, 200]
          "clip-rect": [0, 0, 200, 200]
          "clip-and-scroll": 0
          "backface-visible": true
          type: border
          width: 10
          "border-type": normal
          color: red
          style: dashed
          radius: [180, 180]
        - 
          bounds: [200, 0, 380, 380]
          "clip-rect": [200, 0, 380, 380]
          "clip-and-scroll": 0
          "backface-visible": true
          type: border
          width: 10
          "border-type": normal
          color: green
          style: dashed
          radius: [180, 180]
        - 
          bounds: [580, 0, 200, 200]
          "clip-rect": [580, 0, 200, 200]
          "clip-and-scroll": 0
          "backface-visible": true
          type: border
          width: 10
          "border-type": normal
          color: red  
          style: solid
          radius: [180, 180]
        - 
          bounds: [780, 0, 380, 380]
          "clip-rect": [780, 0, 380, 380]
          "clip-and-scroll": 0
          "backface-visible": true
          type: border
          width: 10
          "border-type": normal
          color: green
          style: solid
          radius: [180, 180]
        - 
          bounds: [0, 380, 200, 200]
          "clip-rect": [0, 380, 200, 200]
          "clip-and-scroll": 0
          "backface-visible": true
          type: border
          width: 10
          "border-type": normal
          color: red
          style: dotted
          radius: [180, 180]
        - 
          bounds: [200, 380, 380, 380]
          "clip-rect": [200, 380, 380, 380]
          "clip-and-scroll": 0
          "backface-visible": true
          type: border
          width: 10
          "border-type": normal
          color: green
          style: dotted
          radius:
                top-left: [180, 180]
                top-right: [0, 0]
                bottom-left: [0, 0]
                bottom-right: [180, 180]
  id: [0, 0]
pipelines: []
```

**Comment by nc4rrillo** [comment]
Looks like one failing reftest:
```
REFTEST TEST-UNEXPECTED-FAIL | reftests/border/border-no-bogus-line.yaml == 
reftests/border/border-no-bogus-line-ref.png | 
image comparison, max difference: 77, number of differing pixels: 476
```

**Comment by glennw** [comment]
Depending on what the differences are, it may just be a case of re-generating the reference image. If you open the ```reftest-analyzer.html``` file you should be able to see the differences.

**Comment by nc4rrillo** [comment]
Compared to the reference the circle looks like it gets a tad bit tighter which shifts everything over a few pixels. May be the result of just how the CSS distribution algorithm works. I will regenerate the test if everything else here looks good.

**Comment by mstange** [comment]
`reftests/border/border-no-bogus-line.yaml` has a border-radius of 40.5 which is slightly larger than half the element's height (80), so it makes sense that it would be affected by these changes.

@subsevenx2001, you added this test in #1975; can you check whether this test still tests what it's supposed to test if this PR lands?

**Comment by mstange** [comment]
The patch looks good to me; it's definitely much easier to understand than the Gecko code.

I would move the calls to `clamp_corner_radius` into the `calculate_ratio` function so that you don't have to pull the radii into local variables twice, and then rename `calculate_ratio` to `ensure_no_corner_overlap` or something like that.

**Comment by subsevenx2001** [comment]
@mstange 
If the reftest result does not show bogus line like the attachment, it is just OK to regenerate the png ref image.
border-no-bogus-line-ref

**Comment by mstange** [comment]
Thanks! I'm just confused why the radius of 40.5 was necessary in the test, instead of just 40. Because Gecko should have limited the radius to 40 on the Gecko side before sending it to WebRender.

**Comment by subsevenx2001** [comment]
@mstange
To choose this strange radius is intended to reproduce #1805 in resolution 1920x1080.
It is choosen by try-and-error.

**Comment by mstange** [comment]
Ok, but with the change in this bug we will treat it as 40.

**Comment by nc4rrillo** [comment]
This change causes a regression with dashed borders:


```
---
root: 
  items: 
    - 
      bounds: [0, 0, 1024, 768]
      "clip-rect": [0, 0, 1024, 768]
      "clip-and-scroll": 0
      "backface-visible": true
      type: "stacking-context"
      "scroll-policy": scrollable
      "transform-style": flat
      filters: []
      items: 
        - 
          bounds: [0, 0, 200, 200]
          "clip-rect": [0, 0, 200, 200]
          "clip-and-scroll": 0
          "backface-visible": true
          type: border
          width: 10
          "border-type": normal
          color: red
          style: dashed
          radius:
                top-left: [180, 180]
                top-right: [0, 0]
                bottom-left: [0, 0]
                bottom-right: [180, 180]
```

**Comment by nc4rrillo** [comment]
Upon further investigation this case also looks glitched on Firefox Nightly with `gfx.webrender.enabled` set to true

[Here's a fiddler](https://jsfiddle.net/LLg2arf9/1/)

**Comment by glennw** [comment]
OK, is this ready for review / merge then?

**Comment by nc4rrillo** [comment]
@glennw it is, I just am not sure what was decided w.r.t regenerating the `border-no-bogus-line` test. 
Code is good to go for review though.

**Comment by mstange** [comment]
I think you can just regenerate the reference. The scenario that I was concerned about is rather theoretical.

**Comment by nc4rrillo** [comment]
Done.

**Comment by glennw** [comment]
Oh, wait - was there a reftest to go with this PR?

**Comment by nc4rrillo** [comment]
I can quick add one to this PR

**Comment by nc4rrillo** [comment]
@glennw done :)

**Comment by glennw** [comment]
Pending Gecko try here:

https://hg.mozilla.org/try/rev/7fbf230e21b8ef4310dfb76fa6db5cf57bb81c56
https://treeherder.mozilla.org/#/jobs?repo=try&revision=7fbf230e21b8ef4310dfb76fa6db5cf57bb81c56

If this is all green when complete, this is ready to merge.

**Comment by glennw** [comment]
Try looks green.

@bors-servo r+ :rocket:

File: webrender/src/border.rs [hunk 1]
```diff
@@ -227,6 +227,48 @@ impl NormalBorderHelpers for NormalBorder {
...
 }
 
+fn ensure_no_corner_overlap(radius: &mut BorderRadius, info: &LayerPrimitiveInfo) {
+    let mut ratio = 1.0;
+    let top_left_radius = &mut radius.top_left;
+    let top_right_radius = &mut radius.top_right;
+    let bottom_right_radius = &mut radius.bottom_right;
+    let bottom_left_radius = &mut radius.bottom_left;
+
+    let sum = top_left_radius.width + bottom_left_radius.width;
```
Comment (mstange) [inline webrender/src/border.rs]: I think this needs to be top_right instead of bottom_left

File: webrender/src/border.rs [hunk 2]
```diff
@@ -227,6 +227,48 @@ impl NormalBorderHelpers for NormalBorder {
...
+    if info.rect.size.width < sum {
+        ratio = f32::min(ratio, info.rect.size.width / sum);
+    }
+
+    let sum = top_right_radius.width + bottom_right_radius.width;
```
Comment (mstange) [inline webrender/src/border.rs]: and this bottom_left instead of top_right

**Pull request merged**